File: R/bert-wordpiece.R
WordPieceTokenizer <- setRefClass(
  "WordPieceTokenizer",
  fields = list(
    vocab = "list",
    unk_token = "character",
    max_word_len = "numeric"
  ),
  methods = list(
    initialize = function(vocab, unk_token = "[UNK]", max_word_len = 100) {
      vocab <<- vocab
      unk_token <<- unk_token
      max_word_len <<- max_word_len
    },
    tokenize = function(text) {
      tokens <- character(0)
      words <- unlist(strsplit(tolower(text), " "))
      for (word in words) {
        word_tokens <- .self$.tokenize_word(word)
        tokens <- c(tokens, word_tokens)
      }
      tokens
    },
    .tokenize_word = function(word) {
      if (nchar(word) > max_word_len) {
        return(c(unk_token))
      }

      output_tokens <- character(0)
      start <- 1
      is_bad <- FALSE

      while (start <= nchar(word)) {
        end <- nchar(word)
        cur_substr <- NULL

        while (start <= end) {
          substr <- substr(word, start, end)
          if (start > 1) {
            substr <- paste0("##", substr)
          }

          if (!is.null(vocab[[substr]])) {
            cur_substr <- substr
            break
          }
          end <- end - 1
        }

        if (is.null(cur_substr)) {
          is_bad <- TRUE
          break
        }

        output_tokens <- c(output_tokens, cur_substr)
        start <- end + 1
      }

      if (is_bad) {
        return(c(unk_token))
      }

      output_tokens
    }
  )
)

File: Julia/bert-wordpiece.jl
mutable struct WordPieceTokenizer
    vocab::Dict{String, Int}
    unk_token::String
    max_word_len::Int
end

function WordPieceTokenizer(vocab::Dict{String, Int}; unk_token::String="[UNK]", max_word_len::Int=100)
    WordPieceTokenizer(vocab, unk_token, max_word_len)
end

function tokenize(tokenizer::WordPieceTokenizer, text::String)
    tokens = String[]
    for word in split(lowercase(text))
        append!(tokens, tokenize_word(tokenizer, word))
    end
    tokens
end

function tokenize_word(tokenizer::WordPieceTokenizer, word::String)
    if length(word) > tokenizer.max_word_len
        return [tokenizer.unk_token]
    end

    output_tokens = String[]
    start = 1
    is_bad = false

    while start <= lastindex(word)
        end_idx = lastindex(word)
        cur_substr = nothing

        while start <= end_idx
            substr = word[start:end_idx]
            if start > 1
                substr = "##" * substr
            end
            if haskey(tokenizer.vocab, substr)
                cur_substr = substr
                break
            end
            end_idx -= 1
        end

        if cur_substr === nothing
            is_bad = true
            break
        end

        push!(output_tokens, cur_substr)
        start = end_idx + 1
    end

    if is_bad
        return [tokenizer.unk_token]
    end

    output_tokens
end
